catalyst_docs/milestone3/milestone3-edgecases.md

```diff
@@ -0,0 +1,295 @@
+        pass
+
+def validator(_: None) -> None:
+    c = MyClass()
+    c.my_method()
+```
+
+### Description:
+
+The above code fails to compile with the error `AssertionError: Invalid Python, class name is undefined at this stage`. The method `my_method` defined inside the class `MyClass` is missing return type but the error message doesn't help the user understand what is wrong with the code.
+
+### Recommendation:
+
+Detect class methods missing return types and throw an explicit error and add a failing test case for this scenario.
+
+## 4. Missing Test Cases for Tuple Assignments
+
+- **Edge case**:
+
+```python
+def validator(a: int) -> int:
+   t1 = (a, a, a)
+   t2 = (a, a)
+
+   t3 = t1 if False else t2
+
+   return t3[2]
+```
+
+### Description:
+
+Though tuples don’t yet have a type syntax (and therefore user-defined functions can’t be created to take tuple arguments), tuples can still be used in other ways that may lead to successful compilation but result in runtime failures, as shown in the example above.
+
+### Recommendation:
+
+In `TupleType.__ge__` in `type_impls.py`, the Python builtin `zip` function is used without checking that the lengths of its arguments are the same. This means a shorter length tuple can potentially be passed into a function whose argument expects a longer length tuple.So ensure the lengths of the TupleTypes are the same when comparing them in `TupleType.__ge__`.
+
+Add test cases for different scenarios related to tuple assignments.
+
+## 5. Test cases for different scenarios of Subtypes in Record
+
+- **File Reference**: `opshin/tests/test_types.py`
+- **Edge Case**:
+
+1. Empty records are not included in the test cases, which leads to a lack of validation for subtyping behavior with empty records.
+
+2. There are no test cases for records with fields in a different order.
+
+3. There are no test cases for records that have the same fields but different constructor IDs.
+
+### Recommendation:
+
+We recommend to add similar tests for the above mentioned edge cases
+
+1. Empty Records:
+
+```
+Empty = RecordType(Record("Empty", "Empty", 0, []))
+```
+
+2. Reversed field order:
+
+```
+A = RecordType(Record("A", "A", 0, [("foo", IntegerInstanceType), ("bar", ByteStringInstanceType)]))
+
+A_reversed = RecordType(Record("A", "A", 0, [("bar", ByteStringInstanceType), ("foo", IntegerInstanceType)]))
+```
+
+3. `Records` having same fields but different constructor ids.
+
+```
+A = RecordType(Record("A", "A", 0, [("foo", IntegerInstanceType), ("bar", ByteStringInstanceType)]))
+
+A_diff_constr = RecordType(Record("A", "A", 1, [("foo", IntegerInstanceType),("bar", ByteStringInstanceType)]))
+
+```
+
+## 6. Uncovered Scenarios in `Union` Type Behavior
+
+- **File Reference**: `opshin/tests/test_Unions.py`
+- **Edge Cases**:
+  The following edge cases were not included in the test suite.
+
+  1.Duplicate entries in `Unions` give compiler errors, but duplicate entries in nested `Unions` don't.
```

Comment:
Let's list the edge cases first and then proceed to the recommendations section. Similar to this:

Edge Cases:
1. ...
2. ...
3. ...

Recommendations:
1. ...
2. ...
3. ...